docs/ldml/tr35-keyboards.md

```diff
@@ -1341,42 +1344,77 @@ _Attribute:_ `modifier` (required for `hardware`)
-> Note that `L` or `R` indicates a left- or right- side modifier only (such as `altL`)
-> whereas `alt` indicates _either_ left or right alt key (that is, `altL` or `altR`). `ctrl` indicates either left or right ctrl key (that is, `ctrlL` or `ctrlR`).
-> `shift` also indicates either shift key. The left and right shift keys are not distinguishable in this specification.
->
-> If there is a layer with a modifier `alt`, there may not be another layer with `altL` or `altR`.  Similarly, if there is a layer with a modifier `ctrl`, there may not be a layer with `ctrlL` or `ctrlR`.
->
-> - `none` (no modifier, may not be combined with others)
-> - `alt`
-> - `altL`
-> - `altR`
-> - `caps`
-> - `ctrl`
-> - `ctrlL`
-> - `ctrlR`
-> - `shift`
->
-> Note that `alt` in this specification is referred to on some platforms as "opt" or "option".
->
-> Left- and right- side modifiers (such as `"altL ctrlR"` or `"altL altR"`) should not be used together in a single `modifier` attribute value.
->
 > For hardware layouts, the use of `@modifier` as an identifier for a layer is sufficient since it is always unique among the set of `layer` elements in a keyboard.
 >
 > The set of modifiers must match `(none|([A-Za-z0-9]+)( [A-Za-z0-9]+)*)`
 >
-> To share a layer between two modifier sets, the layer data must be duplicated.
+> To indicate that no modifiers apply, the reserved name of `none` is used.
+
+**Syntax**
+
+```xml
+<layer id="base"        modifier="none">
+    <row keys="a" />
+</layer>
+
+<layer id="upper"       modifier="shift">
+    <row keys="A" />
+</layer>
+
+<layer id="altgr"       modifier="altR">
+    <row keys="a-umlaut" />
+</layer>
+
+<layer id="upper-altgr" modifier="altR shift">
+    <row keys="A-umlaut" />
+</layer>
+```
+
+#### Layer Modifier Components
+
+ The following modifier components can be used, separated by spaces.
+
+ - `none` (no modifier, may not be combined with others)
+ - `alt`
+ - `altL`
+ - `altR`
+ - `caps`
+ - `ctrl`
+ - `ctrlL`
+ - `ctrlR`
+ - `shift`
+
+(`alt` in this specification is referred to on some platforms as "opt" or "option".)
+
+#### Modifier Left- and Right- keys
+
+1. `L` or `R` indicates a left- or right- side modifier only (such as `altL`)
+ whereas `alt` indicates _either_ left or right alt key (that is, `altL` or `altR`). `ctrl` indicates either left or right ctrl key (that is, `ctrlL` or `ctrlR`).
+
+2.  If there are any layers with a modifier `alt`, there may not also be another layer with `altL` or `altR`.  Similarly, if there is a layer with a modifier `ctrl`, there may not be a layer with `ctrlL` or `ctrlR`.
+
+3. Left- and right- side modifiers may not be mixed together in a single `modifier` attribute value, so neither `altL ctrlR"` nor `altL altR` are allowed.
+
+4. `shift` indicates either shift key. The left and right shift keys are not distinguishable in this specification.
+
+#### Layer Modifier Matching
+
+Layers are matched exactly based on the modifier keys which are down. For example:
```

Comment:
**Problems**

1. The attribute name `modifier` is confusing (as it was in the previous version of keyboards). Please change it to `modifiers`.

The other issues are more serious. I do not think this PR approaches a solution to the problems raised earlier. Without a solution, I don't think we can advance beyond tech preview.

1. The proposed algorithm is still under-specified. In particular, it does not indicate what to do in either of these cases:
    1. None of the layers match. **It is very user-unfriendly to have "gaps" where nothing happens.**
    2. More than one of the layers match.
2. The algorithm and structure are underpowered. For more complex cases (as in **actual keyboards** captured by the previous keyboard structure) one would have to repeat entire layers multiple times. For example, having the caps key down should not affect the result of a control layer; nor an alt layer. But that means duplicating layers in order to have the "crtl caps" and "ctrl", even through the same results obtain.

There are various options to solve these.

1.1 No match
  A. There could be a separate structure that indicates which layer is used if nothing matches.
  B. There could be a special 'modifier' like 'default' that indicates which layer is used.
  C. This relates to a solution for #3

1.2 Multiple matches
  A. Require reading the layers in order.
  B. Declare that such a keyboard is ill-formed, and cannot be interpreted.

2. Underpowered
  A. Revert to the previous keyboard modifier model: modifiers="ctrl caps?"
  B. Introduce an 'or'. That is modifiers="ctrl caps or ctrl" would match either "ctrl caps" or "ctrl". This isn't as concise as the old model, but is trivial to describe and implement.

Reply:
@macchiati the current PR allows multiple modifier sets, so would not require as much repetition.

> For more complex cases (as in actual keyboards captured by the previous keyboard structure) 

These can be helpful for comparison, but it should be noted that it's not a goal to be able to capture or support all actual keyboards.

Reply:
I agree that it is not the goal to support all existing keyboards, but also that it would help adoption if vendors could switch to it for their data.

That said, Windows only needs repeated layers to cover altR = ctrl+alt cases. Does @macchiati have any specific example in mind which would need lots of repetition?